### PR TITLE
gh-85287: Change codecs to raise precise UnicodeEncodeError and UnicodeDecodeError

### PR DESCRIPTION
**Pull request by jjsloboda**
Change codecs to raise `UnicodeEncodeError` and `UnicodeDecodeError`, depending on what operation the API user is attempting, instead of just `UnicodeError`. This involved providing more information to the exceptions, including the problematic string and start and end offsets within the string, which are not always relevant depending on the nature of the exception. 

Some of the codecs will perform multiple encodes and decodes during an operation as part of their implementation. In these cases, I've opted to throw the exception that matches the API function the error is occurring in.

A tricky thing to watch out for while reviewing is that `UnicodeEncodeError` takes a `str` and `UnicodeDecodeError` takes a `bytes`, which is supposed to be the original object being en/decoded. However, most of these codec functions modify their arguments, so by the time the exception block is reached, the original argument is no longer available and needs to be reconstructed. Alternatively, I could change the PR to save the argument, but I don't think it's worth the extra memory just for the exception case (may be more pythonic though).

Another thing to watch for in the review is to make sure I'm not doing anything in the exception blocks that could trigger more en/decoding exceptions when I'm reconstructing the argument strings.

* Issue: gh-85287

### CODE REVIEW

**Comment by davidmerwin** [comment]
>All commit authors signed the Contributor License Agreement. <br/> [CLA signed](https://cpython-clabot.herokuapp.com/signed-contributor-license-agreement?version=96a49432-b8b1-11ec-9bf5-bfe9ad6c72c4)

@cpython-cla-bot @cpython-cla-bot #113701 

Compile and run some source in the interpreter. Arguments are the same as for compile_command(); the default for filename is '<input>', and for symbol is 'single'. One of several things can happen:

The input is incorrect; compile_command() raised an exception (SyntaxError or OverflowError). A syntax traceback will be printed by calling the showsyntaxerror() method. runsource() returns False.
The input is incomplete, and more input is required; compile_command() returned None. runsource() returns True.
The input is complete; compile_command() returned a code object. The code is executed by calling the runcode() (which also handles run-time exceptions, except for SystemExit). runsource() returns False.
The return value can be used to decide whether to use sys.ps1 or sys.ps2 to prompt the next line.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by davidmerwin** [comment]
> ...

@davidmerwin @cpython-cla-bot 

Compile and run some source in the interpreter. Arguments are the same as for compile_command(); the default for filename is '<input>', and for symbol is 'single'. One of several things can happen:

The input is incorrect; compile_command() raised an exception (SyntaxError or OverflowError). A syntax traceback will be printed by calling the showsyntaxerror() method. runsource() returns False.
The input is incomplete, and more input is required; compile_command() returned None. runsource() returns True.
The input is complete; compile_command() returned a code object. The code is executed by calling the runcode() (which also handles run-time exceptions, except for SystemExit). runsource() returns False.
The return value can be used to decide whether to use sys.ps1 or sys.ps2 to prompt the next line.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by jjsloboda** [comment]
Hey @davidmerwin did you mean to comment this somewhere else? I don't see how your comment is relevant to my MR or issue #85287

**Comment by davidmerwin** [comment]
@jjsloboda I am in the process of learning how to collaborate here on GitHub, I apologize if I made a mistake, and thank you for the communication on this

Very Respectfully,
David @davidmerwin

File: Lib/encodings/idna.py
Comment (methane) [inline Lib/encodings/idna.py]:
```suggestion
            raise UnicodeEncodeError("idna", input, 0, 0, f"unsupported error handling {errors}")
```

File: Lib/encodings/idna.py
Comment (methane) [inline Lib/encodings/idna.py]:
```suggestion
                    raise UnicodeEncodeError(
                        "idna", input, index, index+len(label),
                        "label empty or too long")
```

File: Lib/encodings/idna.py
Comment (methane) [inline Lib/encodings/idna.py]:
```suggestion
                raise UnicodeEncodeError("idna", input, index, len(input),
                                         "label too long")
```

File: Lib/encodings/idna.py
Comment (methane) [inline Lib/encodings/idna.py:228]: `len(result)` is not a good offset.

```
>>> "あさ".encode("idna")
b'xn--l8jvb'

>>> "あさ.\u034f".encode("idna")
Traceback (most recent call last):
  File "/Users/inada-n/work/python/cpython/Lib/encodings/idna.py", line 206, in encode
    result.extend(ToASCII(label))
                  ~~~~~~~^^^^^^^
  File "/Users/inada-n/work/python/cpython/Lib/encodings/idna.py", line 93, in ToASCII
    raise UnicodeEncodeError("idna", label, 0, len(label), "label empty or too long")
UnicodeEncodeError: 'idna' codec can't encode characters in position 0--1: label empty or too long

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    "あさ.\u034f".encode("idna")
    ~~~~~~~~~~~~~~~~~~~~^^^^^^^^
  File "/Users/inada-n/work/python/cpython/Lib/encodings/idna.py", line 208, in encode
    raise UnicodeEncodeError(
    ...<5 lines>...
    )
UnicodeEncodeError: 'idna' codec can't encode characters in position 10-9: label empty or too long
encoding with 'idna' codec failed
```

File: Lib/encodings/idna.py
Comment (serhiy-storchaka) [reply Lib/encodings/idna.py]: UnicodeEncodeError implies that there is an error in the encoded string.

Other codecs usually raise LookupError for unknown error handlers, but here an exception is raised even for known error handlers.

Perhaps ValueError better suits here. UnicodeError is a subclass of ValueError, so we can keep it for now.

File: Lib/encodings/idna.py
Comment (serhiy-storchaka) [reply Lib/encodings/idna.py]: Agree we using absolute offset.

But it still reports weird range for empty label. I think that it is better to include an opening or closing dot.

Also, I think that it is better to use different messages for empty and too long labels.

Comment (jjsloboda) [reply Lib/encodings/idna.py]: Split out the empty vs too long cases. Note that `UnicodeEncodeError` takes a python-style [start, end) range, but outputs a (start)-(end-1) inclusive range in the error message, so using something like `0, 0,` gives the weird-looking `0--1` error message whereas `0, 1,` gives `position 0`.

File: Lib/encodings/idna.py
Comment (jjsloboda) [reply Lib/encodings/idna.py]: Changed the cases where there was not an issue with the input string back to `UnicodeError`.

File: Lib/encodings/idna.py
Comment (jjsloboda) [reply Lib/encodings/idna.py:228]: Good call, fixed this to use the lengths of the input seen so far rather than the result string.

File: Lib/encodings/idna.py
Comment (jjsloboda) [reply Lib/encodings/idna.py]: Fixed the linebreaks of some of the longer exceptions to look like this

**Comment by jjsloboda** [comment]
Thanks for the reviews @methane and @serhiy-storchaka, comments addressed and ready for another look

**Review by serhiy-storchaka — COMMENTED** [review]
We cannot simply translate offsets from a label to the full input, because the label is transformed several times during encoding: some characters are ignored, others are combined in the normalization or transformed in the punycode encoding. It is difficult to track offsets of original characters that cannot be encoded. So I suggest to simply report the range of the whole label that cannot be encoded. The main goal is changing the type of the exception to more specific, and it should be enough for now. Later we can narrow the range of error if it is useful and possible.

**Comment by jjsloboda** [comment]
I decided to add some tests for the error offset in one more hope to get the offsets figured out properly, or at least document the current behavior. They seem to be working now at least for those cases, but I'm not an expert in IDNA or Punycode so I'm not sure if there are important edgecases I'm missing.

My goal was to avoid the need to revisit this in the future to further tighten up the offsets, but @serhiy-storchaka let me know if you'd still prefer I switch it to use the whole range of the problematic label.

Ready for another look @methane @serhiy-storchaka

File: Modules/cjkcodecs/multibytecodec.c
Comment (methane) [inline Modules/cjkcodecs/multibytecodec.c]: Do we need copy `inbuf`? If no, use just `inbuf` with `O` format.

File: Modules/cjkcodecs/multibytecodec.c
Comment (methane) [inline Modules/cjkcodecs/multibytecodec.c:835]: I prefer shorter scope.
Declare `excobj` in this scope, and decref manually.

File: Modules/cjkcodecs/multibytecodec.c
Comment (methane) [inline Modules/cjkcodecs/multibytecodec.c]: same here.

File: Modules/cjkcodecs/multibytecodec.c
Comment (methane) [inline Modules/cjkcodecs/multibytecodec.c]: missing `if (excobj == NULL) goto errorexit;`

File: Modules/cjkcodecs/multibytecodec.c
Comment (methane) [inline Modules/cjkcodecs/multibytecodec.c]: same here.

**Comment by nitresor** [comment]
I'm a biginner i just wanna want someone to teach me how we can write the
code in c++

File: Modules/cjkcodecs/multibytecodec.c
Comment (jjsloboda) [reply Modules/cjkcodecs/multibytecodec.c]: Done 👍 I think I tried this originally, but I must have been doing something different because it seems to work fine now.

File: Modules/cjkcodecs/multibytecodec.c
Comment (jjsloboda) [reply Modules/cjkcodecs/multibytecodec.c:835]: Done, for all new and pre-existing code

File: Modules/cjkcodecs/multibytecodec.c
Comment (jjsloboda) [reply Modules/cjkcodecs/multibytecodec.c]: Done, for all new and pre-existing code

File: Modules/cjkcodecs/multibytecodec.c
Comment (methane) [inline Modules/cjkcodecs/multibytecodec.c:958]: pendingsize is size in UTF-8. It is not suitable for `end` value.

File: Modules/cjkcodecs/multibytecodec.c
Comment (methane) [inline Modules/cjkcodecs/multibytecodec.c]: I don't feel this is valid UnicodeEncodeError.
Should we really avoid UnicodeError here?

Comment (jjsloboda) [reply Modules/cjkcodecs/multibytecodec.c]: If `UnicodeEncodeError` is because the user passed an input that directly causes the encoding problem, rather than any problem that occurs during encoding, then `UnicodeError` makes sense here. Changed this one back to `UnicodeError`.

**Comment by jjsloboda** [comment]
Comment addressed @methane , thanks, ready for review

@serhiy-storchaka I think the new IDNA errors work now at least for the main cases, but let me know if you still want me to go back and change the IDNA encoding errors to report the whole label as the problematic range

File: Modules/cjkcodecs/multibytecodec.c
Comment (methane) [inline Modules/cjkcodecs/multibytecodec.c]:
```suggestion
        return NULL;
```

**Comment by methane** [comment]
punycode looks ugly. Although it doesn't have `__all__`, I think only `punycode_decode`, `punycode_encode`, `Codec`, `IncrementalEncoder`, `IncrementalDecoder`, `StreamWriter`, `StreamReader`, and `getregentry` are public APIs.
(Strictly speaking, `getentry` is the only API used from external other than tests.)
So I will change where to encode/decode in them.

**Review by methane — APPROVED** [review]
LGTM. But I will wait for a few weeks to have chance other core developer review my changes.

**Review by methane — APPROVED** [review]

File: Lib/encodings/idna.py
Comment (jjsloboda) [inline Lib/encodings/idna.py:115]: The reason I had a special case for `len(label) == 0` is because if you pass `0, 0` as the range, the error will print the inclusive range as `0--1`, instead of `0-0` for the empty case

Comment (methane) [reply Lib/encodings/idna.py:115]: See "# Step 7: Prepend ACE prefix".
label must not be empty.

**Pull request merged**
